codegen/codegen_test.go

```diff
@@ -0,0 +1,159 @@
+package codegen
+
+import (
+	"context"
+	"strings"
+	"testing"
+
+	"github.com/moby/buildkit/client/llb"
+	"github.com/openllb/hlb/checker"
+	"github.com/openllb/hlb/parser"
+	"github.com/openllb/hlb/solver"
+	"github.com/stretchr/testify/require"
+	"github.com/xlab/treeprint"
+)
+
+func Expect(st llb.State) solver.Request {
+	def, err := st.Marshal(context.Background(), llb.LinuxAmd64)
+	if err != nil {
+		panic(err)
+	}
+
+	return solver.Single(&solver.Params{
+		Def: def,
+	})
+}
+
+type testCase struct {
+	name    string
+	targets []string
+	input   string
+	fn      func(cg *CodeGen) (solver.Request, error)
```

Comment:
similar here, I think we will need to make this `func(t *testing.T, cg *CodeGen) solver.Request` so that we can simplify the error handling and pass the `t` to `Expect`.